File: tests/src/test/javascript/entityauth/EntityAuthenticationDataTest.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Entity authentication data unit tests.
 * 
 * Successful calls to
 * {@link EntityAuthenticationData#create(MslContext, MslObject)} covered in
 * the individual entity authentication data unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("EntityAuthenticationData", function() {
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var MslObject = require('msl-core/io/MslObject.js');
    var EntityAuthenticationData = require('msl-core/entityauth/EntityAuthenticationData.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslError = require('msl-core/MslError.js');
    var MslEntityAuthException = require('msl-core/MslEntityAuthException.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    
    /** Key entity authentication scheme. */
    var KEY_SCHEME = "scheme";
    /** Key entity authentication data. */
    var KEY_AUTHDATA = "authdata";

    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;
    
    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                    result: function(c) { ctx = c; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                initialized = true;
            });
        }
    });
    
    it("no scheme", function() {
        var exception;
        runs(function() {
	        var mo = encoder.createObject();
	        mo.put(KEY_SCHEME + "x", EntityAuthenticationScheme.NONE.name);
	        mo.put(KEY_AUTHDATA, new MslObject());
	        EntityAuthenticationData.parse(ctx, mo, {
	            result: function() {},
	            error: function(e) { exception = e; },
	        });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
    	
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });
    
    it("no authdata", function() {
        var exception;
        runs(function() {
	        var mo = encoder.createObject();
	        mo.put(KEY_SCHEME, EntityAuthenticationScheme.NONE.name);
	        mo.put(KEY_AUTHDATA + "x", new MslObject());
            EntityAuthenticationData.parse(ctx, mo, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });
    
    it("unidentified scheme", function() {
        var exception;
        runs(function() {
	        var mo = encoder.createObject();
	        mo.put(KEY_SCHEME, "x");
	        mo.put(KEY_AUTHDATA, new MslObject());
            EntityAuthenticationData.parse(ctx, mo, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEntityAuthException(MslError.UNIDENTIFIED_ENTITYAUTH_SCHEME));
        });
    });
    
    it("authentication factory not found", function() {
        var ctx;
        runs(function() {
            MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                result: function(c) { ctx = c; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT);

        var exception;
        runs(function() {
            ctx.removeEntityAuthenticationFactory(EntityAuthenticationScheme.NONE);
            var mo = encoder.createObject();
            mo.put(KEY_SCHEME, EntityAuthenticationScheme.NONE.name);
            mo.put(KEY_AUTHDATA, new MslObject());
            EntityAuthenticationData.parse(ctx, mo, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEntityAuthException(MslError.ENTITYAUTH_FACTORY_NOT_FOUND));
        });
    });
});

File: tests/src/test/cpp/entityauth/EntityAuthenticationDataTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <crypto/Key.h>
#include <entityauth/EntityAuthenticationData.h>
#include <entityauth/EntityAuthenticationFactory.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <io/MslEncoderFactory.h>
#include <io/MslObject.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>

#include <gtest/gtest.h>
#include <util/MockMslContext.h>

using netflix::msl::io::MslObject;
using netflix::msl::util::MockMslContext;

using namespace std;
using namespace testing;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

namespace {
/** Key entity authentication scheme. */
const string KEY_SCHEME = "scheme";
/** Key entity authentication data. */
const string KEY_AUTHDATA = "authdata";
} // namespace anonymous

class EntityAuthenticationDataTest : public ::testing::Test
{
public:
	virtual ~EntityAuthenticationDataTest() {}

    EntityAuthenticationDataTest()
    	: ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false))
		, encoder(ctx->getMslEncoderFactory())
	{}

protected:
    shared_ptr<MslContext> ctx;
    shared_ptr<MslEncoderFactory> encoder;
};

TEST_F(EntityAuthenticationDataTest, noScheme)
{
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME + "x", EntityAuthenticationScheme::NONE.name());
    mo->put(KEY_AUTHDATA, make_shared<MslObject>());
    try {
        EntityAuthenticationData::create(ctx, mo);
        ADD_FAILURE() << "should have thrown";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(EntityAuthenticationDataTest, noAuthdata)
{
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, EntityAuthenticationScheme::NONE.name());
    mo->put(KEY_AUTHDATA + "x", make_shared<MslObject>());
    try {
        EntityAuthenticationData::create(ctx, mo);
        ADD_FAILURE() << "should have thrown";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(EntityAuthenticationDataTest, unidentifiedScheme)
{
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, string("x"));
    mo->put(KEY_AUTHDATA, make_shared<MslObject>());
    try {
        EntityAuthenticationData::create(ctx, mo);
        ADD_FAILURE() << "should have thrown";
    } catch (const MslEntityAuthException& e) {
        EXPECT_EQ(MslError::UNIDENTIFIED_ENTITYAUTH_SCHEME, e.getError());
    }
}

TEST_F(EntityAuthenticationDataTest, authFactoryNotFound)
{
	shared_ptr<MockMslContext> ctx = make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false);
	ctx->removeEntityAuthenticationFactory(EntityAuthenticationScheme::NONE);
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_SCHEME, EntityAuthenticationScheme::NONE.name());
    mo->put(KEY_AUTHDATA, make_shared<MslObject>());
    try {
        EntityAuthenticationData::create(ctx, mo);
        ADD_FAILURE() << "should have thrown";
    } catch (const MslEntityAuthException& e) {
        EXPECT_EQ(MslError::ENTITYAUTH_FACTORY_NOT_FOUND, e.getError());
    }
}

}}} // namespace netflix::msl::entityauth
